File: day2/day2.go
package main

import (
	"fmt"
	"math"
	"os"
	"slices"
	"strconv"
	"strings"
)

func checkLineSafe(fields []string) bool {
	max_dist := 3
	safe := true
	inRange := true
	monotonicIncr := true
	monotonicDecr := true
	for i := 0; i < len(fields)-1 && safe; i++ {
		thisNum, _ := strconv.Atoi(fields[i])
		nextNum, _ := strconv.Atoi(fields[i+1])

		if monotonicIncr {
			monotonicIncr = thisNum < nextNum
		}

		if monotonicDecr {
			monotonicDecr = thisNum > nextNum
		}

		if math.Abs(float64(thisNum-nextNum)) > float64(max_dist) {
			inRange = false
		}

		safe = (monotonicDecr || monotonicIncr) && inRange
	}
	return safe
}

func part1(lines []string) {
	sum := 0
	for _, line := range lines {
		if checkLineSafe(strings.Fields(line)) {
			sum++
		}
	}
	fmt.Printf("The number of safe reports in part 1 is: %d\n", sum)
}

func part2(lines []string) {
	sum := 0
	for _, line := range lines {
		fields := strings.Fields(line)
		safe := checkLineSafe(fields)
		for i := 0; i < len(fields) && !safe; i++ {
			// need to assign each time because delete will alter in place
			fields := strings.Fields(line)
			slice := slices.Delete(fields, i, i+1)
			safe = checkLineSafe(slice)
		}
		if safe {
			sum++
		}
	}
	fmt.Printf("The number of safe reports in part 2 is: %d\n", sum)
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Printf("ERROR: %v\n", err)
	}
	lines := strings.Split(string(input), "\n")
	part1(lines)
	part2(lines)
}

File: day2/day2.py
import os

def safe(nums):
    if not nums:
        return False

    valid = True
    is_monotonic_incr = True
    is_monotonic_decr = True

    for i in range(len(nums)-1):
        diff = abs(nums[i] - nums[i+1])
        if not (1 <= diff and diff <= 3):
            valid = False
            break

        if nums[i] < nums[i+1]:
            is_monotonic_incr = False
        elif nums[i] > nums[i+1]:
            is_monotonic_decr = False

    return (is_monotonic_incr or is_monotonic_decr) and valid

def part1(lines):
    sum = 0
    for line in lines:
        if line:
            nums = list(map(int, line.split(" ")))
            if safe(nums):
                sum += 1
    print(sum)

def part2(lines):
    ans = 0
    for line in lines:
        if not line:
            continue

        nums = list(map(int, line.split(" ")))

        if safe(nums):
            ans += 1
        else:
            for i in range(len(nums)):
                subarr = nums.copy()
                subarr.pop(i)
                if safe(subarr):
                    ans += 1
                    break
    print(ans)

with open("input.txt") as inFile:
    lines = inFile.read().split('\n')
    part1(lines)
    part2(lines)

